File: 1370-count-number-of-nice-subarrays/count-number-of-nice-subarrays.java
class Solution {
    public int numberOfSubarrays(int[] nums, int k) {

        HashMap<Integer, Integer> map = new HashMap<>();
        map.put(0, 1);

        int n = nums.length;
        for (int i = 0; i < n; i++) {
            if (nums[i] % 2 == 0) {
                nums[i] = 0;
            } else {
                nums[i] = 1;
            }
        }

        int ans = 0;
        int prefix = 0;
        for (int i = 0; i < n; i++) {
            prefix += nums[i];
            int find = prefix - k;

            if (map.containsKey(find)) {
                ans += map.get(find);
            }

            if (map.containsKey(prefix)) {
                map.put(prefix, map.get(prefix) + 1);
            } else {
                map.put(prefix, 1);
            }
        }
        return ans;
    }
}

File: 1370-count-number-of-nice-subarrays/count-number-of-nice-subarrays.py
class Solution(object):
    def numberOfSubarrays(self, nums, k):
        
        map = {}
        map[0] = 1

        n = len(nums)

        for i in range(n):

            if(nums[i] % 2 == 0):
                nums[i] = 0
            else:
                nums[i] = 1
            
        ans = 0
        prefix = 0

        for ele in nums:

            prefix += ele
            if (prefix-k) in map:
                ans += map[prefix-k]
            
            if prefix in map:
                map[prefix] += 1
            
            else:
                map[prefix] = 1

        return ans